### PR TITLE
Integration with Flyway schema migration in scope of PersistPlugin

### PR DESCRIPTION
**Pull request by belosh59**

### CODE REVIEW

**Comment by belosh59** [comment]
Verified against AWS Services. Build passed. Asset table created. Data filled.
telegram-cloud-photo-size-2-5377853930463146871-y

File: magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/AssetModel.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
+package io.openraven.magpie.plugins.persist;
```
Comment (kickroot) [inline magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/AssetModel.java:1]: Need a copyright header.

**Comment by kickroot** [comment]
Awesome work adding the integration test :+1: 

Looks good to me but I'd love to have @PrzemyslawTusinski look it over as well since he wrote the original persistence code.

**Review by kickroot — APPROVED** [review]

File: magpie-cli/pom.xml [hunk 2]
```diff
@@ -43,26 +43,6 @@
       <artifactId>commons-cli</artifactId>
       <version>1.4</version>
     </dependency>
-    <dependency>
```
Comment (stefansjfw) [inline magpie-cli/pom.xml:46]: nice!

File: magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/FlywayMigrationService.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package io.openraven.magpie.plugins.persist;
```
Comment (stefansjfw) [inline magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/FlywayMigrationService.java:1]: Copyright header here as well

File: magpie-persist/src/test/java/io/openraven/magpie/plugins/persist/PersistPluginIT.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+import static io.openraven.magpie.plugins.persist.TestUtils.getResourceAsString;
+import static org.junit.jupiter.api.Assertions.*;
+
+@ExtendWith(MockitoExtension.class)
+class PersistPluginIT {
+
+  private static final String SELECT_TABLES_META =
+    "SELECT table_name FROM information_schema.tables WHERE table_schema = 'public'";
+  private static final String SELECT_ASSSET_TABLE = "SELECT * FROM assets";
```
Comment (stefansjfw) [inline magpie-persist/src/test/java/io/openraven/magpie/plugins/persist/PersistPluginIT.java:31]: `SELECT_ASSSET_TABLE` -> `SELECT_ASSET_TABLE`

**Review by stefansjfw — APPROVED** [review]

File: magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/AssetModel.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+import java.time.Instant;
+
+public class AssetModel {
```
Comment (PrzemyslawTusinski) [inline magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/AssetModel.java:5]: Why not just use MagpieResource?

File: magpie-persist/src/test/java/io/openraven/magpie/plugins/persist/PersistPluginIT.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  private List<String> selectTables(Handle handle) {
+    return handle
+      .createQuery(SELECT_TABLES_META)
+      .mapTo(String.class)
+      .stream()
+      .collect(Collectors.toList());
+  }
+
+  @Test
+  void whenPersistPluginProcessEnvelopDataSaved() throws Exception {
```
Comment (PrzemyslawTusinski) [inline magpie-persist/src/test/java/io/openraven/magpie/plugins/persist/PersistPluginIT.java:86]: whenPersistPluginProcessEnvelopeDataShouldBeSaved?

File: magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/AssetModel.java (hunk 5 above)
Comment (belosh59) [reply magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/AssetModel.java:5]: I also was thinking about that, coz thats even more simpler .  But I have two concerns:
1. Usage of MagpieResource is bounding with its builder which requires ObjectMapper. 
So in order to create a simple DTO from content I will need to create Object mapper which is not required actually to just simply persist the data.
2. AssetModel in that case associate class naming with table name, so its directly pointing to table structure and table source and doesn't implies any other usages except getting data out of table or saving from that object to table.

If we will use different DTOs on different layers we will achieve less coupling, and will be able to design them respectively to layer requirements (as for example MagpieResource has toJson method providing serialization to ObjectNode, which is not required on persistence layer either).

But if we dont need this, and MagpieResource entity should be treated as asset row I could drop AssetModel. Coz if we will add fields in MagpieResources they will not automatically means that table has such new attributes (tradeoff)

@kickroot @stefansjfw @PrzemyslawTusinski  what do you think?

Comment (kickroot) [reply magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/AssetModel.java:21]: Let's keep it decoupled.

Comment (PrzemyslawTusinski) [reply magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/AssetModel.java:21]: I'd still go with using MagpieResource. 
Every time we will add new field to MagpieResource we will update flyway and upsert method to handle this asset

Comment (kickroot) [reply magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/AssetModel.java:21]: I get that  it is more code, but I'm strongly adverse to having the db schema coupled to the resource definition.  I think the way @belosh59 has done it is the best approach.

File: magpie-persist/src/test/java/io/openraven/magpie/plugins/persist/PersistPluginIT.java (hunk 4 above)
Comment (belosh59) [reply magpie-persist/src/test/java/io/openraven/magpie/plugins/persist/PersistPluginIT.java]: fixed

File: magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/FlywayMigrationService.java (hunk 3 above)
Comment (belosh59) [reply magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/FlywayMigrationService.java:17]: thanks. Fixed

File: magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/AssetModel.java (hunk 1 above)
Comment (belosh59) [reply magpie-persist/src/main/java/io/openraven/magpie/plugins/persist/AssetModel.java:17]: fixed

File: magpie-persist/src/test/java/io/openraven/magpie/plugins/persist/PersistPluginIT.java (hunk 6 above)
Comment (belosh59) [reply magpie-persist/src/test/java/io/openraven/magpie/plugins/persist/PersistPluginIT.java]: thanks. fixed

File: magpie-persist/src/test/java/io/openraven/magpie/plugins/persist/PersistPluginIT.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    persistConfig.setPort(String.valueOf(jdbcDatabaseContainer.getFirstMappedPort()));
+    persistConfig.setUser(jdbcDatabaseContainer.getUsername());
+    persistConfig.setPassword(jdbcDatabaseContainer.getUsername());
+  }
+
+  @Test
+  void whenPersistPluginIntiFlywaySetupSchemaProperly() {
+    // given
+    List<String> tablesBefore = jdbi.withHandle(this::selectTables);
+    assertEquals(2, tablesBefore.size());
```
Comment (stefansjfw) [inline magpie-persist/src/test/java/io/openraven/magpie/plugins/persist/PersistPluginIT.java:65]: @belosh59 Is `2` correct here? Should it be `0`?

**Pull request merged**